### ISSUE TITLE
Go To Definition Not Working

### ISSUE THREAD

**Issue by mukeshdutt** [Microsoft/vscode-go#966]
Hi,
I'm using golang with latest version of vscode and here i'm getting issue **go to definition** but the same thing is working fine in **Sublime Text** editor.

In src folder of my golang workspace here is multiple projects when trying to find same project function, struct etc. those are able to find but when I'm trying **go to definition**  for another project stuff then below message is appeared.

**No definition found for "function name"**.

![screenshot from 2017-05-05 10-15-15](https://cloud.githubusercontent.com/assets/5819408/25734008/75dac1d4-317c-11e7-9169-5455239a568f.png)

**Comment by ramya-rao-a** [comment]
@mukeshdutt 

- Is this a regression? i.e it worked before but not now?
- Just to rephrase what you are saying: go to definition works for functions in the same package but not across multiple packages?
- Have you changed the setting for `go.docsTool`, if yes can you share the value?
- Have you set `go.useLanguageServer` to `true` ?

**Comment by erichiller** [comment]
I too am having this issue, I am unsure which version it began with. It previously did work, but hasn't for ~1-2 months.

Mostly for in-built libraries (runtime, net/http, etc...)

I do not have 
`go.useLanguageServer` set to `true` (it is the default `false`)
Nor have I changed godocs

**Comment by ramya-rao-a** [comment]
@erichiller Can you try setting `go.docsTool` to `gogetdocs` ?

**Comment by bennyscetbun** [comment]
@ramya-rao-a 
Hi since last update, i have the same issue

- It was working fine before the update
- Yes go definition wasnt working between packages
- No i havent changed` go.docsTool`
- No i havent changed `go.useLanguageServer`
- extension: 0.6.61
- I am on MacOS 10.12.4

I just tried to use `gogetdocs` and it works better.

By the way even if my `GOPATH` was set. since last patch i have to set it manually in the vscode configuration.
Same thing for `GOROOT`... (i had an error cannot find go executable in `PATH` or `GOROOT` which is strange...).
Maybe it is related. I was about to make an issue.

**Comment by ramya-rao-a** [comment]
@erichiller It is known issue that `godef` doesn't work well with built-in pkgs. See https://github.com/Microsoft/vscode-go/issues/599 The workaround is to use `gogetdoc` instead.

@bennyscetbun We use `godef` by default for the Go to definition feature. Can you run `godef` from the command line to see if it returns the right location? I cannot think of any reason why it would stop working suddenly.

@bennyscetbun About the GOPATH issue, can you log a separate issue?

**Comment by bennyscetbun** [comment]
@ramya-rao-a
removing `gopath, `goroot` and event `godoctool` from my configuration and restarting all the instance of visual studio fixed all my issues. (i nearly never close my mac session or visual studio...)

**Comment by ramya-rao-a** [comment]
@bennyscetbun great that you are unblocked now. Happy Coding!

@erichiller @mukeshdutt Do you still have issues with the Go to Definition feature?

**Comment by erichiller** [comment]
@ramya-rao-a  Yes - Sorry for the delay - I just changed my config - Restarted vscode as well to include `go.docsTool` to be `gogetdoc`
here are my non-default settings:

```json
  // vscode-go
  // Run 'go build'/'go test -c' on save.
  "go.buildOnSave": false,
  // Run Lint tool on save.
  "go.lintOnSave": true,
  // "go.lintTool": "golint",
  // Run 'go tool vet' on save.
  "go.vetOnSave": true,
  "go.formatOnSave": false,
  "go.gotoSymbol.includeImports": true,
  "go.useCodeSnippetsOnFunctionSuggest": true,
  // Enable gocode's autobuild feature
  "go.gocodeAutoBuild": true,
  // Use gotype on the file currently being edited and report any semantic or syntactic errors found after configured delay.
  "go.liveErrors": {
    "enabled": true,
    "delay": 500
  },
  "go.editorContextMenuCommands": {
    "toggleTestFile": true,
    "addTags": true,
    "removeTags": false,
    "testAtCursor": true,
    "testFile": false,
    "testPackage": false,
    "generateTestForFunction": true,
    "generateTestForFile": true,
    "generateTestForPackage": false,
    "addImport": true,
    "testCoverage": true
  },
  "go.docsTool":"gogetdoc",
  // "go.gopath": "C:\\Users\\ehiller\\dev\\",
  // "go.toolsGopath": "C:\\Users\\ehiller\\dev\\",
  // "go.goroot": "C:\\Users\\ehiller\\AppData\\Local\\omega\\system\\go\\"
  "go.lintTool": "gometalinter",
  "go.lintFlags": [
    "--disable-all",
    // "--enable=varcheck",
    // "--enable=unused",
    // "--enable=ineffassign"
    // "--enable=goconst",
    // "--enable=interfacer",
    // "--enable=staticcheck",
    // "--enable=dupl",
    // "--enable=errcheck",
    // "--enable=golint",
    "--enable=gotype",
    "--enable=vet",
    "--enable=misspell"
    // // "--enable=gocyclo",
    // "--enable=goimports",
    // "--enable=gosimple",
    // "--enable=aligncheck",
    // "--enable=deadcode",
    // "--enable=ineffassign",
    // "--enable=structcheck",
    // // "--enable=unconvert"
  ],
```

**Comment by ramya-rao-a** [comment]
@erichiller So you are saying that "Go to Definition" is not working even after changing to `gogetdoc` ?

**Comment by Emreu** [comment]
Hi everyone! I had possibly same issue.
After some research found root of problem (at least for my case) - function **isPositionInString** (**utils.ts:238**) returns incorrect values for some specific but usual in golang cases. If you have empty string ("") or some bizzare escaped string ("\\\\") before your object of interest - go to definition didn't works.
Examples:
![not_ok_1](https://cloud.githubusercontent.com/assets/601924/26071443/ca5c0d26-39b0-11e7-92c1-eca34f962089.png)
![not_ok_2](https://cloud.githubusercontent.com/assets/601924/26071444/ca621504-39b0-11e7-9599-6dd0dd13fab0.png)
Replace string with something regular and everything works good:
![ok](https://cloud.githubusercontent.com/assets/601924/26071525/0d57d024-39b1-11e7-94bf-a0620d4c22c5.png)

Please consider fixing double quotes counting logic.
Possible solution: count all quotes than subtract escaped quotes.

**Comment by ramya-rao-a** [comment]
@Emreu Thanks for investigating. Next time feel free to submit a PR as well. You have found the root cause, I'd prefer you get the credit for fixing the issue as well :)

For now I have pushed in the fix.

@erichiller @mukeshdutt I'll wait a few more days for you to confirm if your case is similar to @Emreu

**Comment by mukeshdutt** [comment]
@ramya-rao-a Sorry for the delay in my case I'm sending dynamic value(variable) instead of hard coded.

And now i changed **go.docsTool** value to **gogetdoc** after change this gogetdoc plugin was downloaded then again same **go.useLanguageServer** value to true after change the value same as **go.useLanguageServer** plugin tried to download but getting failed message.

**Comment by ramya-rao-a** [comment]
@mukeshdutt

So you are saying setting `go.docsTool` to `gogetdoc` didn't work
And after that you tried, setting `go.useLanguageServer` to `true` also didn't work?

Are you using multiple GOPATHs?

**Comment by hjma29** [comment]
When I use Visual Studio Code with Go exention or Atom with Go-plus, I got "No definition was found for 'Command'" error when trying to check declaration of the following code "&Cobra.Command". JetBrains' Gogland doesn't have this issue and can correctly jump to the Command(a struct) defintion.

https://github.com/spf13/cobra/blob/master/cobra/cmd/add.go#L32

```
var addCmd = &cobra.Command{
    Use:     "add [command name]",
    Aliases: []string{"command"},
    Short:   "Add a command to a Cobra Application",
    Long: `Add (cobra add) will create a new command, with a license and
```

**Comment by ramya-rao-a** [comment]
@hjma29 Can you try setting `go.docsTool` to `gogetdoc`?

**Comment by hjma29** [comment]
Thanks from prompt response. It works right away! Do we have plan to make "gogetdoc" as default or it has other issues comparing with godoc/godef?

**Comment by ramya-rao-a** [comment]
`gogetdoc` tends to be slightly slower than `godoc/godef`. Other than that there are no other issues.

There is no current plan to make it default, but we can try it in one of the updates and see how it fares the larger audience

**Comment by ramya-rao-a** [comment]
We'll use #1025 to track the discussion around making `gogetdoc` the default tool.

@mukeshdutt We haven't heard back from you in a while, hopefully the problem is resolved.

Closing this issue as the fix for the issue @Emreu is in master and will be out in the next update

**Comment by AdiedX** [comment]
I had the same issue with:

VS Code 1.14.1 (June 2017)
Mac OS Sierra 10.12.2
Go Plugin 0.6.62
Go 1.7.5

The `"go.useLanguageServer": true` mentioned by @ramya-rao-a fixed the issue for me.

**Comment by ITSecMedia** [comment]
@AdiedX Thank you, that solved my problem!

I'm using Windows here and **useLanguageServer** is not available but the setting needs to be activated to get the broken "Go to Definition F12" working again while using `"go.docsTool": "gogetdoc",`
